File: jackson/src/main/java/com/github/nstojiljkovic/playSwagger/jackson/JacksonTypeDefinitionGenerator.java
package com.github.nstojiljkovic.playSwagger.jackson;


import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.module.jsonSchema.JsonSchema;
import com.fasterxml.jackson.module.jsonSchema.JsonSchemaGenerator;
import com.fasterxml.jackson.module.jsonSchema.types.ObjectSchema;
import com.github.nstojiljkovic.playSwagger.SwaggerParameterMapper.Parameter;
import com.github.nstojiljkovic.playSwagger.TypeDefinitionGenerator;
import scala.Option;
import scala.collection.JavaConverters;
import scala.collection.Seq;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class JacksonTypeDefinitionGenerator implements TypeDefinitionGenerator {

    final private JsonSchemaGenerator jsonSchemaGenerator;
    final private ObjectMapper mapper;

    public JacksonTypeDefinitionGenerator(final ObjectMapper mapper) {
        this.mapper = mapper;
        this.jsonSchemaGenerator = new JsonSchemaGenerator(mapper);
    }

    private String getTypeName(final JsonSchema schema) {
        switch (schema.getType()) {
            case STRING:
                return "java.lang.String";
            case NUMBER:
                // possible improvement: make differentiation between Double and Float
                return "scala.Double";
            case INTEGER:
                // possible improvement: make differentiation between Long and Int
                return "scala.Long";
            case BOOLEAN:
                return "scala.Boolean";
            case OBJECT:
                if (schema.get$ref() != null) {
                    final String s = schema.get$ref().replace("urn:jsonschema:", "").replaceAll(":", ".");
                    return s;
                } else if (schema.getId() != null) {
                    final String s = schema.getId().replace("urn:jsonschema:", "").replaceAll(":", ".");
                    return s;
                } else {
                    return "scala.Any";
                }
            case ARRAY:
                if (schema.asArraySchema().getItems().isSingleItems()) {
                    JsonSchema itemsSchema = schema.asArraySchema().getItems().asSingleItems().getSchema();
                    return "scala.Seq[" + getTypeName(itemsSchema) + "]";
                } else {
                    // possible improvement: support for multiple item types
                    return "scala.Seq[scala.Any]";
                }
            case NULL:
                return "scala.Null";
            case ANY:
            default:
                return "scala.Any";
        }
    }

    @Override
    public Seq<Parameter> parameters(final Class<?> clazz, final ClassLoader cl) {
        final List<Parameter> parameters = new ArrayList<>();

        try {
            JsonSchema schema = jsonSchemaGenerator.generateSchema(clazz);
            if (schema instanceof ObjectSchema) {
                final Map<String, JsonSchema> properties = ((ObjectSchema) schema).getProperties();
                final Option<String> none = Option.apply(null);
                for (Map.Entry<String, JsonSchema> entry : properties.entrySet()) {
                    final JsonSchema propertySchema = entry.getValue();
                    if (propertySchema == null) {
                        continue;
                    }
                    final Parameter parameter = new Parameter(entry.getKey(), getTypeName(propertySchema), none, none, Option.apply(propertySchema));
                    parameters.add(parameter);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return JavaConverters.asScalaIteratorConverter(parameters.iterator()).asScala().toSeq();
    }
}

File: jackson/src/main/scala/com/github/nstojiljkovic/playSwagger/jackson/JacksonTypeDefinitionGenerator.scala
package com.github.nstojiljkovic.playSwagger.jackson

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.jsonFormatVisitors.JsonFormatTypes
import com.fasterxml.jackson.module.jsonSchema.types.ObjectSchema
import com.fasterxml.jackson.module.jsonSchema.{ JsonSchema, JsonSchemaGenerator }
import com.github.nstojiljkovic.playSwagger.SwaggerParameterMapper.Parameter
import com.github.nstojiljkovic.playSwagger.{ SwaggerParameterMapper, TypeDefinitionGenerator }

import scala.collection.immutable.Seq
import scala.collection.immutable.Map

class JacksonTypeDefinitionGenerator(val mapper: ObjectMapper) extends TypeDefinitionGenerator {
  private val jsonSchemaGenerator: JsonSchemaGenerator = new JsonSchemaGenerator(mapper)

  private def getTypeName(schema: JsonSchema): String = schema.getType match {
    case JsonFormatTypes.STRING =>
      "java.lang.String"
    case JsonFormatTypes.NUMBER =>
      // possible improvement: make differentiation between Double and Float
      "scala.Double"
    case JsonFormatTypes.INTEGER =>
      // possible improvement: make differentiation between Long and Int
      "scala.Long"
    case JsonFormatTypes.BOOLEAN =>
      "scala.Boolean"
    case JsonFormatTypes.OBJECT =>
      if (schema.get$ref != null) {
        val s: String = schema.get$ref.replace("urn:jsonschema:", "").replaceAll(":", ".")
        s
      } else if (schema.getId != null) {
        val s: String = schema.getId.replace("urn:jsonschema:", "").replaceAll(":", ".")
        s
      } else "scala.Any"
    case JsonFormatTypes.ARRAY =>
      if (schema.asArraySchema.getItems.isSingleItems) {
        val itemsSchema: JsonSchema = schema.asArraySchema.getItems.asSingleItems.getSchema
        "scala.Seq[" + getTypeName(itemsSchema) + "]"
      } else {
        // possible improvement: support for multiple item types
        "scala.Seq[scala.Any]"
      }
    case JsonFormatTypes.NULL =>
      "scala.Null"
    case JsonFormatTypes.ANY =>
      "scala.Any"
    case _ =>
      "scala.Any"
  }

  override def parameters(clazz: Class[_])(implicit cl: ClassLoader): Seq[Parameter] = {
    import scala.collection.JavaConverters._

    try {
      val schema: JsonSchema = jsonSchemaGenerator.generateSchema(clazz)
      schema match {
        case s: ObjectSchema =>
          val properties: Map[String, JsonSchema] = Map.empty ++ s.getProperties.asScala
          val none: Option[String] = Option.apply(null)
          Seq.empty ++ properties.toSeq.flatMap(t => {
            val key = t._1
            val propertySchema = t._2
            if (propertySchema != null) {
              Some(SwaggerParameterMapper.Parameter(key, getTypeName(propertySchema), none, none, Option.apply(propertySchema)))
            } else {
              None
            }
          })
        case _ =>
          Seq.empty
      }
    } catch {
      case e: Exception =>
        e.printStackTrace()
        Seq.empty
    }
  }
}
